diff --git a/dwm.c b/dwm.c
index 9972d83..7379a85 100644
--- a/dwm.c
+++ b/dwm.c
@@ -233,7 +233,6 @@ static void clientmessage(XEvent *e);
 static void configure(Client *c);
 static void configurenotify(XEvent *e);
 static void configurerequest(XEvent *e);
-static void copyvalidchars(char *text, char *rawtext);
 static Monitor *createmon(void);
 static void destroynotify(XEvent *e);
 static void detach(Client *c);
@@ -241,6 +240,7 @@ static void detachstack(Client *c);
 static Monitor *dirtomon(int dir);
 static void drawbar(Monitor *m);
 static void drawbars(void);
+static int drawstatusbar(Monitor *m, int bh, int stw, char *text);
 static void enternotify(XEvent *e);
 static void expose(XEvent *e);
 static void focus(Client *c);
@@ -301,6 +301,7 @@ static void showhide(Client *c);
 static void sigchld(int unused);
 static void sigdwmblocks(const Arg *arg);
 static void spawn(const Arg *arg);
+static int status2dtextwidth(const char *text);
 static Monitor *systraytomon(Monitor *m);
 static void spawnbar();
 static void tag(const Arg *arg);
@@ -347,8 +348,8 @@ static Systray *systray =  NULL;
 static const char broken[] = "broken";
 static const char dwmdir[] = "dwm";
 static const char localshare[] = ".local/share";
-static char stext[256];
-static char rawstext[256];
+static char stext[1024];
+static char rawstext[1024];
 static int dwmblockssig;
 pid_t dwmblockspid = 0;
 static int screen;
@@ -590,7 +591,7 @@ buttonpress(XEvent *e)
 			arg.ui = 1 << i;
 		} else if (ev->x < x + TEXTW(selmon->ltsymbol))
 			click = ClkLtSymbol;
-		else if (ev->x > (x = selmon->ww - TEXTW(stext) - getsystraywidth() + lrpad)) {
+		else if (ev->x > (x = selmon->ww - status2dtextwidth(stext) - getsystraywidth() + lrpad)) {
 			click = ClkStatusText;
 			click = ClkStatusText;
 			char *text = rawstext;
@@ -657,7 +658,7 @@ cleanup(void)
 	}
 	for (i = 0; i < CurLast; i++)
 		drw_cur_free(drw, cursor[i]);
-	for (i = 0; i < LENGTH(colors); i++)
+	for (i = 0; i < LENGTH(colors) + 1; i++)
 		free(scheme[i]);
 	free(scheme);
 	XDestroyWindow(dpy, wmcheckwin);
@@ -857,19 +858,6 @@ configurerequest(XEvent *e)
 	XSync(dpy, False);
 }
 
-void
-copyvalidchars(char *text, char *rawtext)
-{
-	int i = -1, j = 0;
-
-	while(rawtext[++i]) {
-		if ((unsigned char)rawtext[i] >= ' ') {
-			text[j++] = rawtext[i];
-		}
-	}
-	text[j] = '\0';
-}
-
 Monitor *
 createmon(void)
 {
@@ -963,6 +951,117 @@ dirtomon(int dir)
 	return m;
 }
 
+int
+status2dtextwidth(const char *text)
+{
+	int i, w = 0;
+	short iscode = 0;
+	char *copy, *p;
+
+	if (!(copy = strdup(text)))
+		die("strdup");
+	p = copy;
+
+	for (i = 0; copy[i]; i++) {
+		if (copy[i] != '^')
+			continue;
+		if (!iscode) {
+			iscode = 1;
+			copy[i] = '\0';
+			w += TEXTW(copy) - lrpad;
+			copy[i] = '^';
+			if (copy[++i] == 'f')
+				w += atoi(copy + ++i);
+		} else {
+			iscode = 0;
+			copy += i + 1;
+			i = -1;
+		}
+	}
+	if (!iscode)
+		w += TEXTW(copy) - lrpad;
+
+	free(p);
+	return w + 2;
+}
+
+int
+drawstatusbar(Monitor *m, int bh, int stw, char *stext)
+{
+	int ret, i, w, x, len;
+	short iscode = 0;
+	char *text, *p;
+
+	len = strlen(stext) + 1;
+	if (!(text = malloc(len)))
+		die("malloc");
+	p = text;
+	memcpy(text, stext, len);
+
+	w = status2dtextwidth(text);
+	ret = x = m->ww - w - stw;
+
+	drw_setscheme(drw, scheme[LENGTH(colors)]);
+	drw->scheme[ColFg] = scheme[SchemeNorm][ColFg];
+	drw->scheme[ColBg] = scheme[SchemeNorm][ColBg];
+	drw_rect(drw, x, 0, w, bh, 1, 1);
+	x++;
+
+	for (i = 0; text[i]; i++) {
+		if (text[i] != '^' || iscode)
+			continue;
+
+		iscode = 1;
+		text[i] = '\0';
+		w = TEXTW(text) - lrpad;
+		drw_text(drw, x, 0, w, bh, 0, text, 0);
+		x += w;
+
+		while (text[++i] && text[i] != '^') {
+			if (text[i] == 'c') {
+				char buf[8];
+				memcpy(buf, text + i + 1, 7);
+				buf[7] = '\0';
+				drw_clr_create(drw, &drw->scheme[ColFg], buf);
+				i += 7;
+			} else if (text[i] == 'b') {
+				char buf[8];
+				memcpy(buf, text + i + 1, 7);
+				buf[7] = '\0';
+				drw_clr_create(drw, &drw->scheme[ColBg], buf);
+				i += 7;
+			} else if (text[i] == 'd') {
+				drw->scheme[ColFg] = scheme[SchemeNorm][ColFg];
+				drw->scheme[ColBg] = scheme[SchemeNorm][ColBg];
+			} else if (text[i] == 'r') {
+				int rx = atoi(text + ++i);
+				while (text[++i] && text[i] != ',');
+				int ry = atoi(text + ++i);
+				while (text[++i] && text[i] != ',');
+				int rw = atoi(text + ++i);
+				while (text[++i] && text[i] != ',');
+				int rh = atoi(text + ++i);
+				drw_rect(drw, rx + x, ry, rw, rh, 1, 0);
+			} else if (text[i] == 'f') {
+				x += atoi(text + ++i);
+			}
+		}
+
+		text += i + 1;
+		i = -1;
+		iscode = 0;
+	}
+
+	if (!iscode) {
+		w = TEXTW(text) - lrpad;
+		drw_text(drw, x, 0, w, bh, 0, text, 0);
+	}
+
+	drw_setscheme(drw, scheme[SchemeNorm]);
+	free(p);
+	return ret;
+}
+
 void
 drawbar(Monitor *m)
 {
@@ -980,9 +1079,7 @@ drawbar(Monitor *m)
 
 	/* draw status first so it can be overdrawn by tags later */
 	if (m == selmon) { /* status is only drawn on selected monitor */
-		drw_setscheme(drw, scheme[SchemeNorm]);
-		tw = TEXTW(stext) - lrpad / 2 + 2; /* 2px extra right padding */
-		drw_text(drw, m->ww - tw - stw, 0, tw, bh, lrpad / 2 - 2, stext, 0);
+		tw = m->ww - drawstatusbar(m, bh, stw, stext);
 	}
 
 	resizebarwin(m);
@@ -2224,7 +2321,8 @@ setup(void)
 	cursor[CurResize] = drw_cur_create(drw, XC_sizing);
 	cursor[CurMove] = drw_cur_create(drw, XC_fleur);
 	/* init appearance */
-	scheme = ecalloc(LENGTH(colors), sizeof(Clr *));
+	scheme = ecalloc(LENGTH(colors) + 1, sizeof(Clr *));
+	scheme[LENGTH(colors)] = drw_scm_create(drw, colors[0], 3);
 	for (i = 0; i < LENGTH(colors); i++)
 		scheme[i] = drw_scm_create(drw, colors[i], 3);
 	/* init system tray */
@@ -2848,7 +2946,8 @@ updatestatus(void)
 	if (!gettextprop(root, XA_WM_NAME, rawstext, sizeof(rawstext)))
 		strcpy(stext, "dwm-"VERSION);
 	else
-		copyvalidchars(stext, rawstext);
+		strncpy(stext, rawstext, sizeof(stext) - 1);
+	stext[sizeof(stext) - 1] = '\0';
 	drawbar(selmon);
 	updatesystray();
 }
@@ -3177,6 +3276,7 @@ xrdb(const Arg *arg)
 {
   loadxrdb();
   int i;
+  scheme[LENGTH(colors)] = drw_scm_create(drw, colors[0], 3);
   for (i = 0; i < LENGTH(colors); i++)
                 scheme[i] = drw_scm_create(drw, colors[i], 3);
   focus(NULL);
